feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/SessionDetailFragment.kt

```diff
@@ -161,6 +170,41 @@ class SessionDetailFragment : DaggerFragment() {
             }
         }
     }
+
+    private fun drawSessionDescription() {
+        val textView = binding.sessionDescription
+        textView.viewTreeObserver.addOnDrawListener {
+            if (textView.lineCount > 5) {
+                if (showEllipsis) {
+                    val end = textView.layout.getLineStart(5)
+                    val ellipsis = getString(R.string.ellipsis_label)
+                    val text = textView.text.subSequence(0, end-ellipsis.length).toString().plus(ellipsis)
```

Comment:
How about this?

```kotlin
                    val text = buildSpannedString {
                        inSpans(object : ClickableSpan() {
                            override fun onClick(widget: View) {
                                val session = binding.session?.desc
                                binding.sessionDescription.text = session
                                showEllipsis = !showEllipsis
                            }

                            override fun updateDrawState(textPaint: TextPaint) {
                                // do nothing
                            }
                        }) {
                            append(textView.text.subSequence(0, end - ellipsis.length))
                            inSpans(object : ForegroundColorSpan(
                                ContextCompat.getColor(
                                    requireContext(),
                                    R.color.colorSecondary
                                )
                            ) {
                                override fun updateDrawState(textPaint: TextPaint) {
                                    textPaint.color = ContextCompat.getColor(
                                        requireContext(),
                                        R.color.colorSecondary
                                    )
                                    textPaint.isUnderlineText = false
                                }
                            }) {
                                append(ellipsis)
                            }
                        }
                    }
```

Reply:
We should remove listener because onDrawListener is called many times 😅 
And onDrawListener can't remove listener inside that listener.
So We can use addOnPreDrawListener. How about this?

```kotlin
    private fun drawSessionDescription() {
        val textView = binding.sessionDescription
        textView.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                val onDrawListener = this
                textView.viewTreeObserver.removeOnPreDrawListener(onDrawListener)
                if (textView.lineCount > 5) {
                    if (showEllipsis) {
                        val end = textView.layout.getLineStart(5)
                        val ellipsis = getString(R.string.ellipsis_label)
                        val text = buildSpannedString {
                            inSpans(object : ClickableSpan() {
                                override fun onClick(widget: View) {
                                    val session = binding.session?.desc
                                    binding.sessionDescription.text = session
                                    showEllipsis = !showEllipsis
                                    textView.viewTreeObserver.addOnPreDrawListener(onDrawListener)
                                }

                                override fun updateDrawState(textPaint: TextPaint) {
                                    // do nothing
                                }
                            }) {
                                append(textView.text.subSequence(0, end - ellipsis.length))
                                inSpans(object : ForegroundColorSpan(
                                    ContextCompat.getColor(
                                        requireContext(),
                                        R.color.colorSecondary
                                    )
                                ) {
                                    override fun updateDrawState(textPaint: TextPaint) {
                                        textPaint.color = ContextCompat.getColor(
                                            requireContext(),
                                            R.color.colorSecondary
                                        )
                                        textPaint.isUnderlineText = false
                                    }
                                }) {
                                    append(ellipsis)
                                }
                            }
                        }
                        binding.sessionDescription.setText(text, TextView.BufferType.SPANNABLE)
                        binding.sessionDescription.movementMethod = LinkMovementMethod.getInstance()
                    }
                }
                return false
            }
        })
    }
```

Reply:
> How about this? I think it is useful to make clickable for all text.
And I used Android-KTX 👍

There is also such a way of writing. informative!

> We should remove listener because onDrawListener is called many times 😅

Surely, Sorry... 🙏